src/dbnode/persist/fs/persist_manager.go

```diff
@@ -654,3 +657,70 @@ func (pm *persistManager) SetRuntimeOptions(value runtime.Options) {
 	pm.currRateLimitOpts = value.PersistRateLimitOptions()
 	pm.Unlock()
 }
+
+var (
+	indexVolumeIndexClaimsLock sync.Mutex
+	indexVolumeIndexClaims     = make(map[string]indexVolumeIndexClaim)
+)
+
+type indexVolumeIndexClaim struct {
+	volumeIndex int
+	blockStart  xtime.UnixNano
+}
+
+// claimNextIndexFileSetVolumeIndex ensures we never use the same volume index
+// no matter what for an index volume from the same process, we always just
+// select the next integer regardless if previous attempt to write to prior
+// index failed or succeeded. This is ok since volume indexes do not need
+// to be contiguous, they just need to be ascending.
+func claimNextIndexFileSetVolumeIndex(
+	opts Options,
+	namespaceMetadata namespace.Metadata,
+	blockStartTime time.Time,
+) (int, error) {
+	indexVolumeIndexClaimsLock.Lock()
+	defer indexVolumeIndexClaimsLock.Unlock()
+
+	filePathPrefix := opts.FilePathPrefix()
+	nowFn := opts.ClockOptions().NowFn()
+
+	// Reap any old entries that are out of retention.
+	retOpts := namespaceMetadata.Options().RetentionOptions()
+	indexOpts := namespaceMetadata.Options().IndexOptions()
+	rp, bs, t := retOpts.RetentionPeriod(), indexOpts.BlockSize(), nowFn()
+	earliestBlockStart := retention.FlushTimeStartForRetentionPeriod(rp, bs, t)
+	earliestBlockStartUnixNanos := xtime.ToUnixNano(earliestBlockStart)
+	for key, claim := range indexVolumeIndexClaims {
+		if claim.blockStart.Before(earliestBlockStartUnixNanos) {
+			delete(indexVolumeIndexClaims, key)
+		}
+	}
+
+	// Now check if previous claim exists.
+	blockStart := xtime.ToUnixNano(blockStartTime)
+	namespace := namespaceMetadata.ID()
+	key := fmt.Sprintf("%s/%s/%d", filePathPrefix, namespace.String(),
```

Comment:
nit: is it better to have a nested map here so we're not generating a ton of keys with duplicated strings for long retention? 

Also, isn't file path prefix static after set? I don't think we have multiple file paths and/or dynamically changing file paths at runtime?